worker/docker.go

```diff
@@ -169,11 +229,58 @@ func (m *DockerManager) HasCapacity(ctx context.Context, pipeline, modelID strin
 		}
 	}
 
+	// TODO: This can be removed if we optimize the selection algorithm.
+	// Currently, using CreateContainer errors only can cause orchestrator reselection.
+	if !m.isImageAvailable(ctx, pipeline, modelID) {
+		return false
+	}
+
 	// Check for available GPU to allocate for a new container for the requested model.
 	_, err := m.allocGPU(ctx)
 	return err == nil
 }
 
+// isImageAvailable checks if the specified image is available locally.
+func (m *DockerManager) isImageAvailable(ctx context.Context, pipeline string, modelID string) bool {
+	imageName, err := m.getContainerImageName(pipeline, modelID)
+	if err != nil {
+		slog.Error(err.Error())
+		return false
+	}
+
+	_, _, err = m.dockerClient.ImageInspectWithRaw(ctx, imageName)
+	if err != nil {
+		slog.Error(fmt.Sprintf("Image for pipeline %s and modelID %s is not available locally: %s", pipeline, modelID, imageName))
+	}
+	return err == nil
+}
+
+// pullImage pulls the specified image from the registry.
+func (m *DockerManager) pullImage(ctx context.Context, imageName string) error {
+	reader, err := m.dockerClient.ImagePull(ctx, imageName, image.PullOptions{})
+	if err != nil {
+		return fmt.Errorf("failed to pull image: %w", err)
+	}
+	defer reader.Close()
+
+	// Show progress.
+	decoder := json.NewDecoder(reader)
+	for {
+		var progress jsonmessage.JSONMessage
+		if err := decoder.Decode(&progress); err == io.EOF {
+			break
+		} else if err != nil {
+			return fmt.Errorf("error decoding progress message: %w", err)
+		}
+		if progress.Status != "" && progress.Progress != nil {
+			fmt.Printf("\r%s: %s", progress.Status, progress.Progress.String())
+		}
+	}
+	fmt.Println()
```

Comment:
Shouldn't we use some logging library here?

Reply:
This is ineed the native approach. Will take a look 👍🏻 .

Reply:
@leszko Implemented to use the logger.